File: Kattis/multigramme kattis.cpp
#include<bits/stdc++.h>
using namespace std;

long int gcd(long int x1,long int x2){
	if(x2!=0){
		return gcd(x2,x1%x2);
	}
	else{
		return x1;
	}
	
}


int main(){
	string s;
	cin>>s;
	long int l,q;
	l=s.size();
	string n="";
	int flag;
	for(int i=0;i<l;i++){
		flag=1;
		for(int j=0;j<n.size();j++){
			if(s[i]==n[j]){
				flag=0;
				break;
			}
		}
		if(flag)n+=s[i];
	}
	q=n.size();
	long int t[q]={0};
	for(int i=0;i<q;i++){
		for(int j=0;j<l;j++){
			if(n[i]==s[j]){
				t[i]++;
			}
		}
	}
	long int g;
	g=t[0];
	for(int i=1;i<q;i++){
		g=gcd(t[i],g);
	}
	cout<<"gcd is "<<g<<endl;
	if(g==1){
		cout<<"-1"<<endl;
	}
	else{
		long int tst;
		for(tst=g;tst>1;tst--){
			if(l%tst==0){
				cout<<"test "<<tst<<endl;
				string b,c="";
				b=s.substr(0,l/tst);
				cout<<b<<endl;
				for(int i=0;i<tst;i++){
					c+=b;
				}
				cout<<"c is: "<<c<<endl;
				if(c==s){
					cout<<b<<endl;
					break;
				}
			}
		}
	}
}

File: Kattis/multigramme kattis.py
def gcd(x1,x2):
    if x2==0:
        return x1
    else:
        return gcd(x2,x1%x2)

ch=input()
N=""
l=len(ch)
for i in range(l):
    if ch[i] not in N:
        N+=ch[i]
q=len(N)
t=[0]*q
for i in range(q):
    for j in range(l):
        if ch[j]==N[i]:
            t[i]+=1
g=t[0]
for i in range(q):
    g=gcd(t[i],g)
if g==1:
    print("-1")
else:
    #print("g is: ",g)
    for tst in range(g,1,-1):
        if l%tst==0:
            o=int(l/tst)
            b=ch[:o]
            #print("b is: ",b)
            ens=set(k for k in b)
            for i in range(1,tst):
                z=ch[i*o:(i+1)*o]
                #print("z is: ",z)
                flag=1
                ens2=set(k for k in z)
                if ens2!=ens:
                    flag=0
                    break
            #print("flag=",flag)
            if flag==1:
                print(b)
                break